File: apps/api/src/WebAppBase.Api/Domain/PasswordPolicy.cs
namespace WebAppBase.Api.Domain;

/// <summary>
/// Mirrors SkaapHond's password rules so a password this app accepts is one SkaapHond
/// will also accept.
/// </summary>
/// <remarks>
/// Kept deliberately identical rather than stricter. Being stricter would reject
/// passwords the auth service allows; being laxer is worse still, because SkaapHond
/// would reject the account and the caller would see a generic failure with no idea
/// which field was wrong.
/// </remarks>
public static class PasswordPolicy
{
    public const int MinimumLength = 8;

    /// <summary>Afrikaans, user-facing: this is shown directly on the registration form.</summary>
    public const string RequirementsMessage =
        "Die wagwoord moet ten minste 8 karakters wees en 'n hoofletter, kleinletter, "
        + "syfer en spesiale karakter bevat.";

    public static bool IsSatisfiedBy(string? password)
    {
        if (string.IsNullOrEmpty(password) || password.Length < MinimumLength)
        {
            return false;
        }

        var hasUpper = false;
        var hasLower = false;
        var hasDigit = false;
        var hasSpecial = false;

        foreach (var character in password)
        {
            if (char.IsUpper(character)) hasUpper = true;
            else if (char.IsLower(character)) hasLower = true;
            else if (char.IsDigit(character)) hasDigit = true;
            else hasSpecial = true;
        }

        return hasUpper && hasLower && hasDigit && hasSpecial;
    }
}

File: apps/nextjs/src/shared/lib/passwordPolicy.ts
/**
 * Mirrors SkaapHond's password rules, and the C# PasswordPolicy that also mirrors them.
 *
 * Kept identical rather than stricter: stricter would reject passwords the auth
 * service allows, laxer would let SkaapHond reject the account after the fact with no
 * indication of which field was wrong.
 */
export const MINIMUM_PASSWORD_LENGTH = 8;

export const PASSWORD_REQUIREMENTS_MESSAGE =
  "Die wagwoord moet ten minste 8 karakters wees en 'n hoofletter, kleinletter, syfer en spesiale karakter bevat.";

export function isPasswordAcceptable(password: string): boolean {
  if (password.length < MINIMUM_PASSWORD_LENGTH) {
    return false;
  }

  return (
    /[A-Z]/.test(password) &&
    /[a-z]/.test(password) &&
    /[0-9]/.test(password) &&
    // Anything that is not a letter or digit counts as special, matching the
    // char.IsUpper/IsLower/IsDigit fallthrough on the server.
    /[^A-Za-z0-9]/.test(password)
  );
}
